### ISSUE TITLE
Zero Grad of w_gate

### ISSUE THREAD

**Issue by panmianzhi** [davidmrau/mixture-of-experts#27]
I implement MoE in the Transformer layer, and I discard the load-balance loss ``l_aux`` during my training. However, I found that the gradient of the w_gate is always zero and that parameter does not update. Is this because I don't consider ``l_aux`` ?

**Comment by panmianzhi** [comment]
OK, I find the reason. The bug is caused by https://github.com/davidmrau/mixture-of-experts/blob/master/moe.py#L245. `top_k_gates = self.softmax(top_k_logits)`. If `k=1` then `top_k_logits` has shape `[batch_size, 1]`, and then the softmax will operate over a single-element axis, which will cause zero-gradient. This issue was discussed in https://stackoverflow.com/questions/76567774/tensorflow-gradients-are-0-even-though-the-loss-is-non-zero.

**Comment by panmianzhi** [comment]
@davidmrau

**Comment by davidmrau** [comment]
did you manage to fix this issue? If so could you make a pull request for it?

**Comment by panmianzhi** [comment]
I think line 242-245 can be modified to
```
logits = F.softmax(logits, dim=1) # newly add
top_logits, top_indices = logits.topk(min(self.k + 1, self.num_experts), dim=1)
top_k_logits = top_logits[:, :self.k]
top_k_indices = top_indices[:, :self.k]
top_k_gates = top_k_logits / (top_k_logits.sum(1, keepdim=True) + 1e-6) # normalization
# top_k_gates = self.softmax(top_k_logits)
```

**Comment by davidmrau** [comment]
Just to understand. What do you mean by discarding the aux loss?

**Comment by panmianzhi** [comment]
I'm using MoE for a different purpose and I'm not currently considering load-balance, so I don't use the loss in https://github.com/davidmrau/mixture-of-experts/blob/master/moe.py#L272C9-L272C67.
In fact this loss may help and I will consider it afterwards.

**Comment by davidmrau** [comment]
Ok I see, did you test whether the code works correctly, after your proposed change?

**Comment by panmianzhi** [comment]
yes, it looks fine on my task  (w/o the auxiliary load-balance loss currently).
BTW, if just replace the origin `top_k_gates = self.softmax(top_k_logits)` with `top_k_gates = top_k_logits / (top_k_logits.sum(1, keepdim=True) + 1e-6)`. Then the code will not work well when `k>1`, because some negative gates value will occur. So I add `logits = F.softmax(logits, dim=1)` at first.

**Comment by panmianzhi** [comment]
I guess if we use the auxiliary load-balance loss, `w_gate` can still be updated based on the origin code when `k=1`. But I think this update is only contributed by the load-balance loss, not the main loss (e.g. the classification loss).
I'm not sure if I'm right and You may need to carefully consider it.
Thank you!

**Comment by davidmrau** [comment]
I'll have a look when I have more time, thanks for your effort in investigating!
